r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs

```diff
@@ -88,54 +260,270 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
 
-    /// Subtract a nonnative field element, without the final reduction step
-    pub fn sub_without_reduce<CS: ConstraintSystemAbstract<ConstraintF>>(
+        // Pad big endian representation to length of normal form
+        let mut per_nonnative_bits = vec![Boolean::Constant(false); len_normal_form - bits.len()];
+        per_nonnative_bits.extend_from_slice(bits);
+
+        // Pack each chunk of `bits_per_limb` many Booleans into a limb,
+        // big endian ordered.
+        let limbs = per_nonnative_bits
+            .chunks_exact(params.bits_per_limb)
+            .enumerate()
+            .map(|(i, chunk)| {
+                // from_bits() assumes big endian vector of bits
+                let limb = FpGadget::<ConstraintF>::from_bits(
+                    cs.ns(|| format!("packing bits to limb {}", i)),
+                    &chunk.to_vec(),
+                )?;
+
+                Ok(limb)
+            })
+            .collect::<Result<Vec<FpGadget<ConstraintF>>, SynthesisError>>()?;
+
+        Ok(NonNativeFieldGadget::<SimulationF, ConstraintF> {
+            limbs,
+            num_of_additions_over_normal_form: BigUint::zero(),
+            simulation_phantom: PhantomData,
+        })
+    }
+
+    /// checks if a NonNativeFieldGadget is odd, i.e. that its mod `p` reduced 
+    /// representation has the least significant bit set to `1`. 
+    #[inline]
+    pub fn is_odd<CS: ConstraintSystemAbstract<ConstraintF>>(
+        &self,
+        mut cs: CS,
+    ) -> Result<Boolean, SynthesisError> {
+        let bits = self.to_bits_strict(cs.ns(|| "to bits strict"))?;
+        Ok(bits[bits.len() - 1])
+    }
+
+
+    /* 
+        arithemtic functions without pre-reduction steps
+    */
+
+    /// Low level function for addition of non-natives. In order to guarantee
+    /// a reducible output, this function assumes that  
+    /// ``
+    ///     bits_per_limb + log(num_add(L) + num_add(R) + 4) <= CAPACITY - 3,
+    /// `` 
+    /// and panics if not.
+    pub(crate) fn add_without_prereduce<CS: ConstraintSystemAbstract<ConstraintF>>(
         &self,
         mut cs: CS,
         other: &Self,
     ) -> Result<Self, SynthesisError> {
+        debug_assert!(
+            self.check() && other.check(),
+            "add_without_prereduce(): check() failed on input gadgets" 
+        );
+
         let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
+        let surfeit = ceil_log_2!(BigUint::from(4usize) + &self.num_of_additions_over_normal_form + &other.num_of_additions_over_normal_form);
+        
+        if params.bits_per_limb + surfeit > ConstraintF::Params::CAPACITY as usize - 3 {
+            return Err(SynthesisError::Other(
+                format!("Security bound exceeded for add_without_prereduce. Max: {}, Actual: {}", 
+                ConstraintF::Params::CAPACITY as usize - 3, 
+                params.bits_per_limb + surfeit))
+            );
+        }
```

Comment:
The reducible condition should be 
```
     bits_per_limb + log(num_add(L) + num_add(R) + 3) <= CAPACITY - 3.
```

Reply:
I think it is right, if you want a reducible sum, then you have to satisfy `bits_per_limb + log(num_add(sum) + 3) <= CAPACITY - 3`, but `num_add(sum) = num_add(L)+num_add(R)+1`, hence overall you get the condition `bits_per_limb + log(num_add(L) + num_add(R) + 4) <= CAPACITY - 3`. So I think that the bound here is correct while in prereduce function `num_add(sum)` is wrongly computed as `num_add(L) + num_add(R)` rather than `num_add(L)+num_add(R)+1`

Reply:
Understood. You are right.